main.cc

```diff
@@ -545,6 +545,84 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
     return linfo.host_id;
 }
 
+static bool raft_topology_change_enabled(const db::config& cfg) {
+    return cfg.consistent_cluster_management() &&
+        cfg.check_experimental(db::experimental_features_t::feature::CONSISTENT_TOPOLOGY_CHANGES);
+}
+
+static void initialize_system_keyspace(sharded<db::system_keyspace>& sys_ks,
+        sharded<cql3::query_processor>& qp,
+        sharded<locator::snitch_ptr>& snitch,
+        distributed<replica::database>& db,
+        sharded<locator::effective_replication_map_factory>& erm_factory,
+        const db::config& cfg,
+        sharded<gms::feature_service>& feature_service)
+{
+    supervisor::notify("starting system keyspace");
+    sys_ks.start(std::ref(qp), std::ref(db), std::ref(snitch)).get();
+    // TODO: stop()?
+
+    // Initialization of a keyspace is done by shard 0 only. For system
+    // keyspace, the procedure  will go through the hardcoded column
+    // families, and in each of them, it will load the sstables for all
+    // shards using distributed database object.
+    // Iteration through column family directory for sstable loading is
+    // done only by shard 0, so we'll no longer face race conditions as
+    // described here: https://github.com/scylladb/scylla/issues/1014
+    supervisor::notify("loading system sstables");
+    replica::distributed_loader::init_system_keyspace(sys_ks, erm_factory, db, system_table_load_phase::phase1).get();
+
+    // Re-enable previously enabled features on node startup.
+    // This should be done before commitlog starts replaying
+    // since some features affect storage.
+    feature_service.local().enable_features_on_startup(sys_ks.local(), raft_topology_change_enabled(cfg)).get();
```

Comment:
Reeeeally? The initialize_system_keyspace() is supposed to initialize system keyspace. Well, maybe init schema commitlog :P (below). Enabling features looks completely alien here

Reply:
What can we do? Don't extract `initialize_system_keyspace`? It's like if I don't see the problem maybe it'll go away. We can bury this dependency between features and `system_keyspace` boot load in the sea of `scylla_main` and nobody will notice it until the next code movements which will break it.

Reply:
But feature_service is not system keyspace, it depends on it. Everything we do later in main depends on system keyspace, but you don't put it in  initialize_system_keyspace().

Reply:
> Don't extract initialize_system_keyspace? 

Yes. The whole main() really consists of 3 logical blocks
- configure seastar app
- start sharded services
- do necessary initialization steps
- ... then call join_cluster()

All three are intermixed :( The 3rd is the largest. What this patch does is peeks random pieces from it and puts them into a separate function (of a strange name). Not that I consider this code to be at good shape, but this particular change doesn't make it better

Reply:
> But feature_service is not system keyspace, it depends on it. Everything we do later in main depends on system keyspace, but you don't put it in initialize_system_keyspace().

1. We put into `initialize_system_keyspace` things related to `system_keyspace` initialization;
2. We put `enable_features_on_startup` in it because we have two way dependency, `system_keyspace` depends on features and features depend on `system_keyspace`;
3. This circular dependency exists regardless of whether we extract `initialize_system_keyspace` or not. The separate function just highlights the problem.

Reply:
I see.

Reply:
> All three are intermixed :( The 3rd is the largest. 

ok, then why not split it into functions to make the code manageable, like we do in all other places, where a single function grows too big? I think the code should be like this:
```
sharded<Service1> service1;
sharded<Service2> service2;
...
void initialize_service1() {
  service1.start(...);
  // other necessary load//boot/wiring stuff
}

void initialize_service2() {
...
}

void initialize_services() {
  initialize_service1();
  initialize_service2();
  ...
}
```
We can put shutdown actions in a container and run them in the appropriate place, e.g. at the end of `scylla_main`. We can even create a simple factory helper
```
template<T, Args...>
void init_with_stop(sharded<T>& service, Args... args) {
  service.start(args...);
  if constexpr(T has stop method) {
    shutdown_actions.push_back([]() { service.stop(); });
  }
}
```
and get rid of ~90% of `defer_verbose_shutdown` boilerplate

But this can't be done in one go, the code is too convoluted, that's why I 'peek random pieces from it and puts them into a separate function'. I don't know, what's strange about the name 'initialize' though.

Reply:
@gusev-p your suggestion will break the rollback ordering between services called with init_with_stop() and random objects on the frame. We can fix it by making init_with_stop() return an `undo` object.

```c++
undo stop_blah = init_with_stop(blah, arg1, arg2);
```

Reply:
Here are some ways to solve the feature_service <-> system_keyspace dependency loop

1. store features outside the database (BAD)
2. read features from sstables + mutation_reader directly
3. create a system_keyspace object with instructions to only read the features table, construct a private database + query_processor on top, read features, destroy these temporary objects, initialize features with result of previous read, continue (maximum reuse but a long road to get there)

This problem is also seen in filesystems which store their own metadata in files.

Reply:
> your suggestion will break the rollback ordering between services called with init_with_stop() and random objects on the frame. We can fix it by making init_with_stop() return an undo object.

If we had simple initialisation code structure like 
```
void initialize_services() {
  initialize_service1();
  initialize_service2();
  ...
}
```
we wouldn't have random objects on the frame.

This doesn't preclude extracting the `initialize_` functions though, we can return the destroy object from it and store it on the stack.

Reply:
> ok, then why not split it into functions to make the code manageable, like we do in all other places

I read this question as "why don't we make main being structured already". I don't know how to answer it, sorry :) I can only rant about this place was much worse in this sense 4 years ago and there's still some route to be walked with it. You're welcome to join, but from my POV _this_ particular change doesn't make it better.

> I don't know, what's strange about the name 'initialize' though.

Because this brand new `initialize_system_keyspace()` does much more than what it name says. I disagree that enabling features on startup is the part of system keyspace initialization. The fact that system keyspace cannot be initialized without enabled features still doesn't justify it. Some other systems also need enabled features and they thus piggy-back system keyspace initialization thus not contributing to "structured and manageable code" concept.

Reply:
FYI, there's another ... bag of accumulated not-immediately-obviously-structured (but apparently mostly working) code out there. Check the `storage_service::join_cluster()` -> `storage_service::join_token_ring()` (-[partially]-> `storage_service::bootstrap()`) methods.

Reply:
> I don't know, what's strange about the name 'initialize' though.

Maybe we can approach this from slightly different angle. If it's indeed the initialization of the system keyspace, then try making this code be a (preferably non-static) method of the system_keyspace class and place it in db/system_keyspace.cc. _That_ would make it obvious if the change makes the code more structured or less structured.

Reply:
> I mean, why not start splitting scylla_main into functions right now?

I'm all for it, but the `initialize_system_keyspace` doesn't fit. As I wrote somewhere else in this PR, turning this function into a system_keyspace's method -- likely does.

> There are two points in this discussion:
>   *  should we extract initialize_system_keyspace_thread or not;

Provided this function is system_keyspace's method -- yes, we should try to. Provided it's not -- no we shouldn't

> I can make it an instance method of system_keyspace, but it doesn't feel right to me. I'm more for some scylla_configurator class, which contains all the wiring logic and can be reused in boost tests, for example.

Yes, we have a related issue: #2795. If the configurator you're talking about in question would suite both -- main() and single_node_cql_test_env::do_with() -- I'll likely change my mind about `initialize_system_keyspace`

Reply:
> If we had simple initialisation code structure like
> 
>     void initialize_services() {
>       initialize_service1();
>       initialize_service2();
>       ...
>     }

From what I've seen so far, that's very likely utopian goal. We can definitely try to arrange `sharded<service>::start`-s and `invoke_on_all(&service::start)`-s in a plain sequence, but some initialization parts will still be intermixed between different services. Recent invasion of group0_<whatever> (and related) in main proves that. Three ways I can see how to overcome the latter.

1. untie it with either inter-dependent semaphores/condvars/on_...-notifications (like you did here).
2. split services into smaller parts introducing more sharded<Foo> stuff in main
3. relax and make the "bottom-half" part of the initialization explicitly intermixed

Reply:
> Yes, we have a related issue: https://github.com/scylladb/scylladb/issues/2795. If the configurator you're talking about in question would suite both -- main() and single_node_cql_test_env::do_with() -- I'll likely change my mind about initialize_system_keyspace

I can create a new class in a separate file, `scylla_service_host` (sounds more appropriate), move my new functions into it, along with the services they care about, reuse this class in `scylla_main` and `cql_test_env.cc`. This will set the direction for the refactoring. It is impossible to do it from start to finish in one PR.

Reply:
Do I understand your suggestion right:
  * revert my changes about extracting functions from `scylla_main` in this PR;
  * try to do the same in another PR, but with your implication that these changes will not be successful and will most likely be reverted too. ("you'll likely see that this brand new scylla_service_host thing is just becoming that very main() renamed and moved to another file.")

This doesn't work on a motivational level. I can only revert the changes and never touch this topic again.

Reply:
> revert my changes about extracting functions from scylla_main in this PR;

Not quite. I accept the `initialize_local_info_thread` change and only suggest not to introduce `initialize_system_keyspace` _here_, because this change grows out of "remove flushes when writing to system tables" scope. As  you told above: "It is impossible to do it from start to finish in one PR."

> try to do the same in another PR, but with your implication that these changes will not be successful and will most likely be reverted too. ("you'll likely see that this brand new scylla_service_host thing is just becoming that very main() renamed and moved to another file.")

I didn't say it would not be successful. I just shared my thoughts that structuring main is more than moving its pieces into functions with "descriptive" names; and my expectation that this code move would end up being just code move.

> This doesn't work on a motivational level. I can only revert the changes and never touch this topic again.

I feel your pain, some of my changes were rejected by maintainers as well. Apparently every time a change is submitted, it can be asked to be reworked or even thrown away. There's even the explicit "request changes" option in the "submit my review" form here on github.

Reply:
> only suggest not to introduce initialize_system_keyspace here, because this change grows out of "remove flushes when writing to system tables" scope. As you told above: "It is impossible to do it from start to finish in one PR."

The essence of the change was not extracting `initialize_system_keyspace` by itself, but all the preparatory work - gather its pieces from various parts of `scylla_main` and making them work together - splitting `sstables_format_selector`, introducing `on_system_tables_loaded` interface, inlining `system_keyspace.setup` parts into the correct places. Those changes were dictated by their own reasons, but in the end we got something relatively cohesive, and now we are just refusing to put a name on it for some reason.

If we want to structure `scylla_main`, then the work above is what we have to do to many other services - make their initialisation routines as independent from each other as possible and define interfaces between them where necessary. I agree with you that this is more than 'moving pieces and giving names'.

Reply:
> The essence of the change was not extracting `initialize_system_keyspace` by itself, but all the preparatory work - gather its pieces from various parts of `scylla_main` and making them work together - splitting `sstables_format_selector`, introducing `on_system_tables_loaded` interface, inlining `system_keyspace.setup` parts into the correct places. Those changes were dictated by their own reasons, but in the end we got something relatively cohesive, and now we are just refusing to put a name on it for some reason.

So let's merge those preparatory, gathering, making-them-work-together, splitting, introducing, etc. changes! Just keep them in one bundle inside main.cc _or_ place in the system_keyspace::initialize_on_boot_or_whatever() non-static method to show that it's indeed self-contained. Moving those into main-local static method is code move for the sake of the code move itself. The amount of arguments you need to carry along and templatized list of notify-me services speaks for itself from my perspective.

> If we want to structure `scylla_main`, then the work above is what we have to do to many other services - make their initialisation routines as independent from each other as possible and define interfaces between them where necessary. I agree with you that this is more than 'moving pieces and giving names'.

:+1:

Reply:
ok, I'll leave it 'as is' inside `scylla_main`, moving it into `system_keyspace` just creates too many strange dependencies to it (e.g. `supervisor::notify`)

Reply:
But I really don't get the arguments here, they sound arbitrary to me.

I think the code of the form
```
initialize_service1();
initialize_service2();
...
initialize_service158();
```
is better than then same code where all the functions are inlined. At least we can use function names instead of line numbers when we reason about the code.

Reply:
It's better, but this patch doesn't make it so. The change is

```patch
@@ -10,9 +10,7
     mess_a
     mess_b
     mess_c
-    mess_d
-    mess_e
-    mess_f
+    initialize_foo()
     mess_g
     mess_h
     mess_i
@ -20,6 +20,11
  
  
  
+void initialize_foo() {
+    mess_d
+    mess_e
+    mess_f
+}
  
  
  
```

you say: "it's a step in right direction"
I say: "I doubt it, but let's better try to make at least two or, better, three such steps in a separate PR"

Reply:
ok